### PR TITLE
feat: use precompiles

### PR DESCRIPTION
**Pull request by 0xWOLAND**
Uses precompiles for BLS12381 acceleration

### CODE REVIEW

File: src/scalar.rs [hunk 1]
```diff
@@ -390,15 +390,15 @@ impl Scalar {
...
-        }
+        // cfg_if! {
+        //     if #[cfg(target_os = "zkvm")] {
+        //         let mut res = *self;
+        //         res.mul_inp(self);
+        //         res
+        //     } else {
+        self._square()
+        //     }
+        // }
```
Comment (ratankaliani) [inline src/scalar.rs]: nit: Remove these comments.

File: src/fp6.rs [hunk 2]
```diff
@@ -270,12 +270,48 @@ impl Fp6 {
...
+                    tmp.add_inp(&self.c2);
+                    let mut c2 = b.c0;
+                    c2.add_inp(&b.c2);
+                    c2.mul_inp(&tmp);
+                    tmp = t0;
+                    tmp.add_inp(&t2);
+                    c2.sub_inp(&tmp);
+                    c2.add_inp(&t1);
+                    Fp6 { c0, c1, c2 }
+                } else {
```
Comment (ratankaliani) [inline src/fp6.rs]: I'd add a more detailed comment for what's going on here. Especially given that many people will be using this library it's important to be clear about how you got this (e.g. addchain).

File: src/fp2.rs [hunk 3]
```diff
@@ -553,29 +568,29 @@ impl Fp2 {
...
+            let byte_vec = read_vec();
+            let bytes: [u8; 97] = byte_vec.try_into().unwrap();
+            match bytes[96] {
+                0 => CtOption::new(Fp2::zero(), Choice::from(0u8)),
+                _ => {
+                    let inv = Fp2::from_bytes(&bytes[0..96].try_into().unwrap()).unwrap();
+                    CtOption::new(inv, !self.is_zero() & (self * inv).ct_eq(&Fp2::one()))
+                }
+            }
+        }
```
Comment (ratankaliani) [inline src/fp2.rs]: Add a more detailed comment for why you're doing `_invert`. Might be better to call it `cpu_invert`.

File: src/fp.rs [hunk 4]
```diff
@@ -362,7 +362,7 @@ impl Fp {
         // works for elements that are actually quadratic residue,
         // so we check that we got the correct result at the end.
 
-        let sqrt = self.pow_vartime(&[
+        let sqrt = self.pow_vartime_unconstrained(&[
```
Comment (ratankaliani) [inline src/fp.rs]: Explain why you're doing this in a comment.

File: src/fp6.rs (hunk 2 above)
Comment (0xWOLAND) [reply src/fp6.rs]: Screenshot 2024-08-19 at 5 09 00 PM

Oh I just directly implemented this algorithm verbatim

**Pull request merged**